README.md
# react-native-searchbar

A pretty awesome search bar for React Native. It has a built in simple search capability which can be utilised by handing the component `data` and providing a `handleResults` function which will receive the filtered copy of the `data` array.

The search function uses a depth-first algorithm to traverse the data and simply looks for an `indexOf` the user input within all strings, numbers, and boolean values in the data.

Works on both iOS and Android.

![react-native-searchbar](http://i.imgur.com/i21YwnW.gif)

(Rendered with the example in these docs and repository)

## Installation

- `npm install react-native-searchbar-controlled --save`
- Install [`react-native-vector-icons`](https://github.com/oblador/react-native-vector-icons) if the project doesn't have them already. The search bar accesses MaterialIcons.
- Now you can require the search bar with `import SearchBar from 'react-native-searchbar-controlled'` or `var SearchBar = require('react-native-searchbar-controlled')`

## Available Props

| Property                      | Type      | Default              | Description                                                                                     |
| :---------------------------- | :-------- | :------------------- | :---------------------------------------------------------------------------------------------- |
| input                         | string    | `''`                 | The value of the searchbox                                                                      |
| placeholder                   | string    | `Search`             | The placeholder for the search bar.                                                             |
| handleChangeText(input)       | function  | `null`               | Fires when the input text changes.                                                              |
| onSubmitEditing               | function  | `null`               | Fired after pressing 'search' on the device keyboard                                            |
| onHide()                      | function  | `null`               | Fires when the search bar is hidden.                                                            |
| onBack                        | function  | Hides the search bar | Fires when the back icon is pressed.                                                            |
| onX                           | function  | `null`               | Fires when the X button is pressed. Still clears the input.                                     |
| onFocus                       | function  | `null`               | Fires when the search bar receives focus.                                                       |
| onBlur                        | function  | `null`               | Fires when the search bar loses focus.                                                          |
| backButton                    | Component | required             | Sets the back button component.                                                                 |
| backButtonAccessibilityLabel  | string    | `Navigate up`        | Accessibility label for the back button.                                                        |
| closeButton                   | Component | required             | Sets the close button component.                                                                |
| closeButtonAccessibilityLabel | string    | `Clear search text`  | Accessibility label for the close button.                                                       |
| backCloseSize                 | number    | 28                   | Sets the size of back button & close button.                                                    |
| heightAdjust                  | number    | `0`                  | Adjust the height of the search bar.                                                            |
| backgroundColor               | string    | `white`              | The background colour of the search bar.                                                        |
| iconColor                     | string    | `gray`               | The color of the back and X icons.                                                              |
| textColor                     | string    | `gray`               | The color of the search bar text.                                                               |
| selectionColor                | string    | `lightskyblue`       | The color of the the search bar cursor and text selection.                                      |
| placeholderTextColor          | string    | `lightgray`          | The color of the placeholder text.                                                              |
| animate                       | boolean   | `true`               | Animate the search bar when it is shown and hidden.                                             |
| animationDuration             | number    | `200`                | The duration of the above animation in milliseconds.                                            |
| showOnLoad                    | boolean   | `false`              | Show the search bar when it is first rendered.                                                  |
| hideBack                      | boolean   | `false`              | Hide the back button.                                                                           |
| hideX                         | boolean   | `false`              | Hide the X icon which clears the input.                                                         |
| iOSPadding                    | boolean   | `true`               | Pad the size of the iOS status bar.                                                             |
| iOSHideShadow                 | boolean   | `false`              | Hide the shadow under the search bar in iOS.                                                    |
| iOSPaddingBackgroundColor     | string    | `transparent`        | Change the background color of the padding.                                                     |
| clearOnShow                   | boolean   | `false`              | Clear input when the search bar is shown.                                                       |
| clearOnHide                   | boolean   | `true`               | Clear input when the search bar is hidden.                                                      |
| clearOnBlur                   | boolean   | `false`              | Clear input when the search bar is blurred.                                                     |
| focusOnLayout                 | boolean   | `true`               | Focus the text input box whenever it is shown.                                                  |
| autoCorrect                   | boolean   | `true`               | AutoCorrect the search bar text.                                                                |
| autoCapitalize                | string    | `sentences`          | Auto capitialize behaviour of the text input - enum('none', 'sentences', 'words', 'characters') |
| keyboardAppearance            | string    | 'default'            | Determines the color of the keyboard.                                                           |
| fontFamily                    | string    | `System`             | The font family to be used.                                                                     |
| fontSize                      | number    | 20                   | Sets the font size.                                                                             |
| allDataOnEmptySearch          | boolean   | `false`              | Search results behave as a `.filter`, returning all data when the input is an empty string.     |

## Usage

- Use a ref to show and hide the search bar and set the text input value `ref={(ref) => this.searchBar = ref}`
- `this.searchBar.show()`
- `this.searchBar.hide()`

##### Notes for Android

- Render the search bar component after the component it is supposed to display over. iOS handles this nicely with a `zIndex` of `10`. Android elevation is set to 2.
- The bottom of the search bar will have a thin border instead of a shadow.

## Example

Full example at `example/`

```javascript
import SearchBar from 'react-native-searchbar';

const items = [
  1337,
  'janeway',
  {
    lots: 'of',
    different: {
      types: 0,
      data: false,
      that: {
        can: {
          be: {
            quite: {
              complex: {
                hidden: [ 'gold!' ],
              },
            },
          },
        },
      },
    },
  },
  [ 4, 2, 'tree' ],
];

...
_handleResults(results) {
  this.setState({ results });
}
...

...
<SearchBar
  ref={(ref) => this.searchBar = ref}
  data={items}
  handleResults={this._handleResults}
  showOnLoad
/>
...
```

## Contributing

Contributing to `react-native-searchbar` is easy! With four simple steps:

### Create a branch

1. Fork the repository
1. `git clone <your-repo-url>` to clone your GitHub repo to your local one
1. `git pull origin master` to pull the latest code
1. `npm install` to install the project's dependencies
1. `git checkout -b the-name-of-my-branch` to create a branch (use something short and comprehensible, such as: `fix-styling-of-search-bar`).

### Make the change

### Test the change

1. Run `npm run fix` from the project root (This will run Prettier and ESLint and automatically fix any issues).
1. If possible, test any visual changes in Android and iOS.

### Push the change!

1. `git add -A && git commit -m "My message"` (replacing `My message` with a commit message, such as `Fixed styling on search bar`) to stage and commit your changes
1. `git push my-fork-name the-name-of-my-branch`


CHANGELOG.md
# v2.0.0
- Removed internal searching code
- Made icons required from the outside
- Removed lodash and react-native-vector-icons dependencies

# v1.0.0
- Initial release


index.js
import React, { Component } from 'react';
import {
  StyleSheet,
  Dimensions,
  Platform,
  View,
  TextInput,
  TouchableOpacity,
  Animated,
} from 'react-native';
import PropTypes from 'prop-types';

const INITIAL_TOP = Platform.OS === 'ios' ? -80 : -60;

export default class Search extends Component {
  static propTypes = {
    input: PropTypes.string.isRequired,
    placeholder: PropTypes.string,
    handleChangeText: PropTypes.func.isRequired,
    onSubmitEditing: PropTypes.func,
    onFocus: PropTypes.func,
    onBlur: PropTypes.func,
    onHide: PropTypes.func,
    onBack: PropTypes.func,
    onX: PropTypes.func,
    backButton: PropTypes.object.isRequired,
    backButtonAccessibilityLabel: PropTypes.string,
    closeButton: PropTypes.object.isRequired,
    closeButtonAccessibilityLabel: PropTypes.string,
    backCloseSize: PropTypes.number,
    fontSize: PropTypes.number,
    heightAdjust: PropTypes.number,
    backgroundColor: PropTypes.string,
    iconColor: PropTypes.string,
    textColor: PropTypes.string,
    selectionColor: PropTypes.string,
    placeholderTextColor: PropTypes.string,
    animate: PropTypes.bool,
    animationDuration: PropTypes.number,
    showOnLoad: PropTypes.bool,
    hideBack: PropTypes.bool,
    hideX: PropTypes.bool,
    iOSPadding: PropTypes.bool,
    iOSPaddingBackgroundColor: PropTypes.string,
    iOSHideShadow: PropTypes.bool,
    clearOnShow: PropTypes.bool,
    clearOnHide: PropTypes.bool,
    clearOnBlur: PropTypes.bool,
    focusOnLayout: PropTypes.bool,
    autoCorrect: PropTypes.bool,
    autoCapitalize: PropTypes.string,
    keyboardAppearance: PropTypes.string,
    fontFamily: PropTypes.string,
    editable: PropTypes.bool,
  };

  static defaultProps = {
    input: '',
    placeholder: 'Search',
    backButtonAccessibilityLabel: 'Navigate up',
    closeButtonAccessibilityLabel: 'Clear search text',
    heightAdjust: 0,
    backgroundColor: 'white',
    iconColor: 'gray',
    textColor: 'gray',
    selectionColor: 'lightskyblue',
    placeholderTextColor: 'lightgray',
    animate: true,
    animationDuration: 200,
    showOnLoad: false,
    hideBack: false,
    hideX: false,
    iOSPadding: true,
    iOSPaddingBackgroundColor: 'transparent',
    iOSHideShadow: false,
    clearOnShow: false,
    clearOnHide: true,
    clearOnBlur: false,
    focusOnLayout: true,
    autoCorrect: true,
    autoCapitalize: 'sentences',
    keyboardAppearance: 'default',
    fontFamily: 'System',
    backCloseSize: 28,
    fontSize: 20,
    editable: true,
  };

  constructor(props) {
    super(props);
    this.state = {
      show: props.showOnLoad,
      top: new Animated.Value(
        props.showOnLoad ? 0 : INITIAL_TOP + props.heightAdjust
      ),
    };
  }

  show = () => {
    const {
      animate,
      animationDuration,
      clearOnShow,
      handleChangeText,
    } = this.props;
    if (clearOnShow) {
      handleChangeText('');
    }
    this.setState({ show: true });
    if (animate) {
      Animated.timing(this.state.top, {
        toValue: 0,
        duration: animationDuration,
        useNativeDriver: true,
      }).start();
    } else {
      this.setState({ top: new Animated.Value(0) });
    }
  };

  hide = () => {
    const { onHide, animate, animationDuration } = this.props;
    if (onHide) {
      onHide();
    }
    if (animate) {
      Animated.timing(this.state.top, {
        toValue: INITIAL_TOP,
        duration: animationDuration,
        useNativeDriver: true,
      }).start();
      const timerId = setTimeout(() => {
        this._doHide();
        clearTimeout(timerId);
      }, animationDuration);
    } else {
      this.setState({ top: new Animated.Value(INITIAL_TOP) });
      this._doHide();
    }
  };

  _doHide = () => {
    const { clearOnHide, handleChangeText } = this.props;
    this.setState({ show: false });
    if (clearOnHide) {
      handleChangeText('');
    }
  };

  _handleX = () => {
    const { onX } = this.props;
    this._clearInput();
    if (onX) onX();
  };

  _handleBlur = () => {
    const { onBlur, clearOnBlur } = this.props;
    if (onBlur) {
      onBlur();
    }
    if (clearOnBlur) {
      this._clearInput();
    }
  };

  _clearInput = () => {
    this._onChangeText('');
  };

  _onChangeText = input => {
    this.props.handleChangeText(input);
  };

  render = () => {
    const {
      placeholder,
      heightAdjust,
      backgroundColor,
      iconColor,
      textColor,
      selectionColor,
      placeholderTextColor,
      onBack,
      hideBack,
      hideX,
      iOSPadding,
      iOSPaddingBackgroundColor,
      iOSHideShadow,
      onSubmitEditing,
      onFocus,
      focusOnLayout,
      autoCorrect,
      autoCapitalize,
      keyboardAppearance,
      fontFamily,
      backButton,
      backButtonAccessibilityLabel,
      closeButton,
      closeButtonAccessibilityLabel,
      backCloseSize,
      fontSize,
      editable,
    } = this.props;

    return (
      <Animated.View
        style={[
          styles.container,
          {
            transform: [{ translateY: this.state.top }],
            shadowOpacity: iOSHideShadow ? 0 : 0.7,
          },
        ]}
      >
        {this.state.show && (
          <View style={[styles.navWrapper, { backgroundColor }]}>
            {Platform.OS === 'ios' && iOSPadding && (
              <View
                style={{
                  height: 20,
                  backgroundColor: iOSPaddingBackgroundColor,
                }}
              />
            )}
            <View
              style={[
                styles.nav,
                { height: (Platform.OS === 'ios' ? 52 : 62) + heightAdjust },
              ]}
            >
              {!hideBack && backButton && (
                <TouchableOpacity
                  accessible={true}
                  accessibilityComponentType="button"
                  accessibilityLabel={backButtonAccessibilityLabel}
                  onPress={onBack || this.hide}
                >
                  <View style={{ width: backCloseSize, height: backCloseSize }}>
                    {backButton}
                  </View>
                </TouchableOpacity>
              )}
              <TextInput
                ref={ref => (this.textInput = ref)}
                onLayout={() => focusOnLayout && this.textInput.focus()}
                style={[
                  styles.input,
                  {
                    fontSize: fontSize,
                    color: textColor,
                    fontFamily: fontFamily,
                    marginLeft: hideBack ? 30 : 0,
                    marginTop:
                      Platform.OS === 'ios' ? heightAdjust / 2 + 10 : 0,
                  },
                ]}
                selectionColor={selectionColor}
                onChangeText={input => this._onChangeText(input)}
                onSubmitEditing={() =>
                  onSubmitEditing ? onSubmitEditing() : null
                }
                onFocus={() => (onFocus ? onFocus() : null)}
                onBlur={this._handleBlur}
                placeholder={placeholder}
                placeholderTextColor={placeholderTextColor}
                value={this.props.input}
                underlineColorAndroid="transparent"
                returnKeyType="search"
                autoCorrect={autoCorrect}
                autoCapitalize={autoCapitalize}
                keyboardAppearance={keyboardAppearance}
                editable={editable}
              />
              {closeButton && (
                <TouchableOpacity
                  accessible={true}
                  accessibilityComponentType="button"
                  accessibilityLabel={closeButtonAccessibilityLabel}
                  onPress={
                    hideX || this.props.input === '' ? null : this._handleX
                  }
                >
                  <View style={{ width: backCloseSize, height: backCloseSize }}>
                    {closeButton}
                  </View>
                </TouchableOpacity>
              )}
            </View>
          </View>
        )}
      </Animated.View>
    );
  };
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    zIndex: 10,
    position: 'absolute',
    elevation: 2,
    shadowRadius: 5,
  },
  navWrapper: {
    width: Dimensions.get('window').width,
  },
  nav: {
    ...Platform.select({
      android: {
        borderBottomColor: 'lightgray',
        borderBottomWidth: StyleSheet.hairlineWidth,
      },
    }),
    flex: 1,
    flexBasis: 1,
    flexDirection: 'row',
    justifyContent: 'space-around',
    alignItems: 'center',
  },
  input: {
    ...Platform.select({
      ios: { height: 30 },
      android: { height: 50 },
    }),
    width: Dimensions.get('window').width - 120,
  },
});
